### ISSUE TITLE
normalize order of tokens in tasks

### ISSUE THREAD

**Issue by edio** [mpcjanssen/simpletask-android#175]
That allows tasks to look consistently. The order of tokens should be configurable.

For example, if user defines the order as follows 

```
$date $description $tags $lists $threshold $due
```

then the following entries

```
2012-12-12 Buy milk @shopping due:2012-12-13 t:2012-12-12 +groceries 
2012-12-12 +vacation @online Order tickets due:2012-12-29
```

should look like

```
2012-12-12 Buy milk +groceries @shopping t:2012-12-12 due:2012-12-13
2012-12-12 Order tickets +vacation @online due:2012-12-29
```

2 approaches:
1. parse task on save and reorder tokens before flushing data to `FileStoreInterface`
2. reorder only when displaying
#### Normalize on save

Pros:
- All new tasks are normalized
- Display logic remains the same simple
- No confusion when editing task because you see exactly what you have in file

Cons:
- Old tasks should be updated or will remain inconsistent
- Possible inconsistency when edited manually with text editor
- Parsing on save is required
#### Normalize on display

Pros:
- All tasks (old and new) look normalized
- No need to parse on save

Cons:
- Display logic become much more complex
- Encourages chaos in todo.txt

**Comment by smichel17** [comment]
I think we should do as little normalization on display as possible. The more processing, the less useful the plain text file is without the app, and portability is a big draw of todo.txt. For making tasks easily skimmable, I'd encourage syntax highlighting instead.

I also don't like normalization on write. If a user wants to add both `+buy milk @store` and `go to @store and +buy milk` to their todo list, I think we should let them.

I propose that the real problems (and their solutions) are:

1. The current order is hard to skim
  - Continue with 'Normalization on insert'
  - [ ] switch to `$prio $date $threshold $due $description $tags $lists`. Constant length fields at start.
    - easier to skim.
    - while editing, date & threshold can be inserted the same way as priority, keeping the cursor in the same position.
    - I don't have strong feelings about whether threshold or due should come first. I do care about the other values.
2. It's not easy to deal with a messy list.
  - [ ] Add a 'normalize' function that can be used on a task or group of tasks.

**Comment by mpcjanssen** [comment]
Will not implement this in code. If there are lua hooks you can build this yourself.

**Comment by smichel17** [comment]
>Will not implement this in code. If there are lua hooks you can build this yourself. 

Do you mean a normalized view, a normalize function, or both? If both, you could close this issue.

**Comment by mpcjanssen** [comment]
Both as separate hooks

**Comment by smichel17** [comment]
+1 for lua hooks. I'll note that the desktop todo app I use has the ability to normalize lists & tags on display and I don't find it nearly as useful as Simpletask's heading.

**Comment by mpcjanssen** [comment]
Can be implemented using the existing onAdd and the upcoming onUpdate Lua hooks.

**Comment by mpcjanssen** [comment]
Fixed with #933.
